adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/PropertyBag.java

```diff
@@ -0,0 +1,99 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.adapter.mqtt;
+
+import io.netty.handler.codec.http.QueryStringDecoder;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A collection of methods for processing <em>property-bag</em> set at the end of a topic.
+ *
+ */
+public final class PropertyBag {
+
+    private final Map<String, List<String>> properties;
+    private final String topic;
+    private final String topicWithoutPropertyBag;
+
+    private PropertyBag(final String topic) {
+        this.topic = topic;
+        this.properties = getProperties();
+        this.topicWithoutPropertyBag = extractTopicWithoutPropertyBag();
+    }
+
+    /**
+     * Creates a property bag object from the given topic.
+     * 
+     * @param topic The topic that the message has been published to.
+     * @return The property bag object.
+     * @throws NullPointerException if topic is {@code null}.
+     */
+    public static PropertyBag fromTopic(final String topic) {
+        Objects.requireNonNull(topic);
+        return new PropertyBag(topic);
+    }
+
+    /**
+     * Gets a property value from the <em>property-bag</em>.
+     *
+     * @param name The property name.
+     * @return The property value or {@code null} if the property is not set.
+     */
+    public String getProperty(final String name) {
+        return Optional.ofNullable(properties)
+                .map(props -> props.get(name))
+                .map(values -> values.get(0))
+                .orElse(null);
+    }    
+
+    /**
+     * Returns the topic without the <em>property-bag</em>.
+     *
+     * @return The topic without the <em>property-bag</em>.
+     */
+    public String topicWithoutPropertyBag() {
+        return topicWithoutPropertyBag;
+    }
+
+    /**
+     * Trims the <em>property-bag</em> from the topic and returns the rest.
+     *
+     * @return The topic without the <em>property-bag</em>.
+     */
+    private String extractTopicWithoutPropertyBag() {
+        final int index = topic.lastIndexOf("/?");
```

Comment:
IMHO we should do this and *getProperties* in the same method/step in order to prevent searching for the index twice ...

Reply:
@sophokles73 I have pushed the changes, pls have a look again.